xds/src/main/java/io/grpc/xds/CdsLoadBalancer.java

```diff
@@ -230,6 +232,34 @@ public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
     }
   }
 
+  @VisibleForTesting
+  static class EdsLoadBalancingHelper extends ForwardingLoadBalancerHelper {
+    private final Helper delegate;
+    private final UpstreamTlsContext upstreamTlsContext;
+
+    EdsLoadBalancingHelper(Helper helper, UpstreamTlsContext upstreamTlsContext) {
+      this.delegate = helper;
+      this.upstreamTlsContext = upstreamTlsContext;
```

Comment:
Should this really be final?

Reply:
Maybe an `AtomicReference<UpstreamTlsContext>` so that it can be updated?

Reply:
If you see where this is called (line 297), a new instance of `EdsLoadBalancingHelper` is created whenever a new LoadBalancer is created and takes the delegate helper and `UpstreamTlsContext` as arguments to the constructor. There are multiple instances as needed. Okay to make them final.

Reply:
`onClusterChanged()` can be called multiple times. Each time it may contain a different `UpstreamTlsContext`, but your helper is only created for the first time. So I suggested using `AtomicReference<UpstreamTlsContext>`.

Reply:
Oh okay, good point. But then this needs a bigger change than just using  `AtomicReference<UpstreamTlsContext>`. Specifically when `edsBalancer != null` (line 292) I need to access the `Helper` inside the existing `edsBalancer`. Isn't that right?

Reply:
One way is to declare the helper in `ClusterWatcherImpl` and init it in the constructor

```java
final EdsLoadBalancingHelper helper;
ClusterWatcherImpl(Helper helper, ResolvedAddresses resolvedAddresses) {
      this.helper = new EdsLoadBalancingHelper(helper, new AtomicReference<>());
      this.resolvedAddresses = resolvedAddresses;
}
```

Reply:
@dapengzhang0 after talking to @ejona86 I looked at the code and I need to set the `Attributes` on each EAG in the list inside `CreateSubchannelArgs`. I made that change and also took care of null as we discussed.